src/Microsoft.ML.Data/Transforms/HashTransform.cs

```diff
@@ -555,13 +588,23 @@ private ValueGetter<VBuffer<uint>> ComposeGetterVec(IRow input, int iinfo)
                     if (len > 0 && src.Length != len)
                         throw Host.Except("Hash transform expected {0} slots, but got {1}", len, src.Length);
 
+                    T[] values = src.Values;
+                    // force-densify the input in case hasherSparse is not provided
```

Comment:
Because it densifies *before* we hash, which forces the same hash to be done for every input -- we have to do that I think for "ordered" inputs, but "ordered" inputs are not the default.

So: maybe we could add an internal condition to test the orderness and if so have a more efficient code in that case.

So imagine an (unordered) hash happening on a sparse float vector of length 1 million and count 2, we could imagine 0 being hashed once (even being cached by capture in the delegate!), then copied into the array, which will be far more efficient.